### ISSUE TITLE
Make PUT and POST operations for repositories Ansible friendly

### ISSUE THREAD

**Issue by sephiroth-j** [DependencyTrack/dependency-track#822]
The PUT and POST operations to `/v1/repository` are unsuitable for automated management with Ansible (_as example_). PUT completely ignores the "uuid" from the body and generates a new one upon successful creation. POST ignores changes of "identifier" although "uuid" being the unique identifier. With this behavior, it is not possible to manage the configuration using Ansible. It would be necessary to DELETE a repo first using its uuid and then create it with PUT. But in this case, the uuid will change and it must be stored in some sort of state-file. That is not how Ansible is meant to be used.

### Proposed Behavior:
- allow changes of "identifier" using POST
- PUT should respect "uuid" from the request body

**Comment by stevespringett** [comment]
All PUTs in the entirety of all APIs intentionally ignore UUIDs provided by clients. This is by design and ensures that only the persistence layer can generate them, ensures there are no collisions, and that they were generated using UUID v4.

This is a common design pattern and should be supported by the majority of management apps.

**Comment by sephiroth-j** [comment]
Well, you wrote it, the persistence layer enforces uniqueness - using an appropriate unique key constraint. It does not know who generated the value - it could also be enforced and guaranteed if it is a user-provided value. The API endpoint could and should check that the value matches a given format - for example using appropriate JSR-380 annotations.

What is the point of having a request model that requires a certain field, but which itself is completely ignored? That is an unexpected and undocumented behavior. On the other hand, the field "identifier" is optional but has a significant meaning - it is shown on the UI and can not be updated. In other words: it should be required for the PUT request in opposite to "uuid".

> This is a common design pattern and should be supported by the majority of management apps.

Infrastructure as code describe the state a system (_or particular part of it_) should have and creates/restores this state - it does not hold information about the current/previous state. Of course it would be possible to delete all unexpected repositories in the first step and create only the expected ones in the second. But again, this will not work because "uuid" is a required input field which is ignored and changes afterwards. So the state which Dependency-Track should have (_limited to this particular scope of its configuration_) is unreachable.

**Comment by stevespringett** [comment]
> What is the point of having a request model that requires a certain field, but which itself is completely ignored? 

you're likely referring to the fact the autogenerated Swagger docs do not represent reality. This is a known issue and the only solution is to eliminate autogenerated docs and replace them with manual doc authoring. The swagger describes fields that the endpoint does not accept. Likewise the docs do not include parameters that should be included. 

I think you're mis-interpreting my statement with regard to the persistence layer. JDO has a customValueStrategy on UUID fields for every persistable object. This completely offloads the creation of UUIDs to Datanucleus. No code exists in DT that creates or updates UUIDs.

https://github.com/DependencyTrack/dependency-track/blob/92acd751742cb1f06754502381e5bfc6ac535d05/src/main/java/org/dependencytrack/model/Repository.java#L87

Have you pinged the Slack channel? There's a lot of users with ephemeral infrastructure that use DT.

Also, I'm a bit confused regarding the use case. Dependency-Track does not support the embedded H2 database. Per the docs, it's included for demonstration and evaluation purposes only. All DT instances should be using a permanent external database.

**Comment by sephiroth-j** [comment]
> you're likely referring to the fact the autogenerated Swagger docs do not represent reality. This is a known issue and the only solution is to eliminate autogenerated docs and replace them with manual doc authoring. The swagger describes fields that the endpoint does not accept. Likewise the docs do not include parameters that should be included.

Yes, you are right. I think I did not tried the PUT request without uuid and solely relied on the swagger docs.

> I think you're mis-interpreting my statement with regard to the persistence layer. JDO has a customValueStrategy on UUID fields for every persistable object. This completely offloads the creation of UUIDs to Datanucleus. No code exists in DT that creates or updates UUIDs.
> 
> https://github.com/DependencyTrack/dependency-track/blob/92acd751742cb1f06754502381e5bfc6ac535d05/src/main/java/org/dependencytrack/model/Repository.java#L87

Ah, okay, I understand. I thought there exists own code that would create the UUID (e.g. constructor based generated value).

> Also, I'm a bit confused regarding the use case. Dependency-Track does not support the embedded H2 database. Per the docs, it's included for demonstration and evaluation purposes only. All DT instances should be using a permanent external database.

The use case is to manage one or more (durable) DT instances with Ansbile. We do not use H2 for DT - I also never mentioned H2.

I guess I was just confused by the misleading Swagger docs and thought that my proposal would be a good enhancement when using IaC for DT.

What about the changeability of the "identifier" field?

**Comment by stevespringett** [comment]
Understood. I do think there's room for improvement in the repository API around changing the 'identifier' field. The persistence layer enforces uniqueness based on the identifier and repository type. So you couldn't have more than one maven repo named 'maven central' for example. Currently the code ignores if the user POSTs an update to a repo and changes the identifier. This could be changed (with a bit of cross-database testing) to support changing the identifier.

**Comment by sephiroth-j** [comment]
> The persistence layer enforces uniqueness based on the identifier and repository type. 
That's interesting. I would have expected that the repository URL must be unique but not the name (identifier). Instead, the URL does not have to be unique or conform to a valid schema.

**Comment by jrevillard** [comment]
Hello @sephiroth-j , @stevespringett ,

Would you be aware if somebody already created an ansible module for DT ?

Best,
Jerome

**Comment by valentijnscholten** [comment]
What could help here is if projects would have an `externalID` or something like that. That way ansible and other scripts can have a  self controlled value to use without affecting the inner workings of DT. I know there are project properties already, but they are a bit "hidden" and I think not searchable, so you can't find projects "having property A being set to value Y".
